### DISCUSSION TITLE
How to extend spacing from inside a plugin?

### DISCUSSION THREAD

**Discussion by unrevised6419** [tailwindlabs/tailwindcss#9342] [Help] [answered]
It is possible with any of `add*` or `match*` functions from within a plugin to extend config `spacing`?

Or how correctly to extend anything from inside a plugin?

My use case is that I have a custom container (not core) that changes `padding-inline` depending on screen size.

```css
@screen screen-name {
  .container {
    padding-inline: ##px;
  }
}
```

Now I added a utility that cancels that padding when needed using negative `margin-inline`

```css
@screen screen-name {
  .container-cancel {
    margin-inline: -##px;
  }
}
```

But now I need to use `.container-cancel` depending on screen size. For example a column on mobile is full width and I cancel container padding, so there are no gutters, but on desktop, it should not do that.

And I thought maybe I can add container padding as a spacing unit, and I can use it like this.

```html
<div class="container">
  <div>some content with paddings</div>
  <div class="-m-container-gutter lg:m-0">some content with paddings only on desktop</div>
</div>
```

For the moment I have it like this with one more class `.container-cancel-none`

```ts
const containerPaddings = plugin.withOptions((paddings) => (options) => {
	const { addComponents, config } = options;
	const screens = config('theme.screens');
	const screenKeys = Object.keys(screens);

	addComponents(
		screenKeys.map((name) => ({
			[`@screen ${name}`]: {
				'.container': {
					'padding-inline': paddings[name],
				},
				'.container-cancel': {
					'margin-inline': `-${paddings[name]}`,
				},
			},
		})),
	);
	addComponents({
		'.container-cancel-none': {
			'margin-inline': 0,
		},
	});
});
```

**Comment by wongjn** [comment, marked as answer]
> It is possible with any of `add*` or `match*` functions from within a plugin to extend config `spacing`?

No, by the point that any of these functions are run, the config has all been resolved.

> Or how correctly to extend anything from inside a plugin?  

Yep! You can return a Tailwind config with a [second function argument to `plugin.withOptions()`](https://tailwindcss.com/docs/plugins#exposing-options):

```js
plugin.withOptions(
  (paddings) => (options) => {
    // …
  },
  (options) => ({
    theme: {
      extend: {
        // …
      },
    },
  }),
);
```
---

For your use case, you could add your gutter as a CSS variable. This avoids having the `.container-cancel` in a media query, and so you should then be able to use variants like `md:container-cancel` or override it with the `.container-cancel-none` like `container-cancel md:container-cancel-none`.

```js
const containerPaddings = plugin.withOptions((paddings) => (options) => {
	const { addComponents, addBase, addUtilities, config } = options;
	const screens = config('theme.screens');
	const screenKeys = Object.keys(screens);

	addBase(
		screenKeys.map((name) => ({
			[`@screen ${name}`]: {
				':root': {
					'--tw-gutter': paddings[name],
				},
			},
		})),
	);

	addComponents({
		'.container': {
			'padding-inline': 'var(--tw-gutter, 0px)',
		},
	});

	addUtilities({
		'.container-cancel': {
			'margin-inline': 'calc(var(--tw-gutter, 0px) * -1)',
		},
	});

	addUtilities({
		'.container-cancel-none': {
			'margin-inline': 0,
		},
	});
});
```

Or you could look at extending the Tailwind config for padding/margin/spacing, as per stated previously.

**Reply by unrevised6419** [reply]
Thanks a lot! 🙌
